src/platform/ESP32/DiagnosticDataProviderImpl.cpp

```diff
@@ -287,6 +287,64 @@ void DiagnosticDataProviderImpl::ReleaseNetworkInterfaces(NetworkInterface * net
     }
 }
 
+CHIP_ERROR DiagnosticDataProviderImpl::GetThreadMetrics(ThreadMetrics ** threadMetricsOut)
+{
+#ifdef CONFIG_FREERTOS_USE_TRACE_FACILITY
+    ThreadMetrics * head = nullptr;
+    uint32_t arraySize   = uxTaskGetNumberOfTasks();
+
+    Platform::ScopedMemoryBuffer<TaskStatus_t> taskStatusArray;
+    VerifyOrReturnError(taskStatusArray.Calloc(arraySize), CHIP_ERROR_NO_MEMORY);
+
+    uint32_t dummyRunTimeCounter;
+    arraySize = uxTaskGetSystemState(taskStatusArray.Get(), arraySize, &dummyRunTimeCounter);
+
+    for (uint32_t i = 0; i < arraySize; i++)
+    {
+        auto thread = static_cast<ThreadMetrics *>(Platform::MemoryCalloc(1, sizeof(ThreadMetrics)));
+        if (thread == nullptr)
+        {
+            // Clean up previously allocated nodes
+            while (head != nullptr)
+            {
+                auto temp = head->Next;
+                Platform::MemoryFree(head);
+                head = temp;
+            }
+            return CHIP_ERROR_NO_MEMORY;
+        }
+
+        Platform::CopyString(thread->NameBuf, taskStatusArray[i].pcTaskName);
+        thread->name.Emplace(CharSpan::fromCharString(thread->NameBuf));
+        thread->id = taskStatusArray[i].xTaskNumber;
+        thread->stackFreeMinimum.Emplace(taskStatusArray[i].usStackHighWaterMark);
+
+        thread->stackFreeCurrent.ClearValue();
+        thread->stackSize.ClearValue();
```

Comment:
Please check if we can use `pxStackBase`, `pxTopOfStack`, and `pxEndOfStack` members of the `TaskStatus_t` structure to calculate these values.